### PR TITLE
hypervisor: mshv: Add support to handle GPA Attribute Intercept

### PR DESCRIPTION
**Pull request by jinankjain**
Microsoft Hypervisor raises this request to change the guest page visibility in case of SEV-SNP guest.

### CODE REVIEW

**Comment by jinankjain** [comment]
@russell-islam @likebreath PTAL :)

**Comment by likebreath** [comment]
@russell-islam The changes are mainly MSHV exclusive. Can you please take a look first?

File: hypervisor/src/mshv/mod.rs [hunk 1]
```diff
@@ -606,6 +608,56 @@ impl cpu::Vcpu for MshvVcpu {
 
                     Ok(cpu::VmExit::Ignore)
                 }
+                #[cfg(feature = "sev_snp")]
+                hv_message_type_HVMSG_GPA_ATTRIBUTE_INTERCEPT => {
+                    let info = x.to_gpa_attribute_info().unwrap();
+                    let host_vis = info.__bindgen_anon_1.host_visibility();
+                    if host_vis >= HV_MAP_GPA_READABLE | HV_MAP_GPA_WRITABLE {
+                        warn!("Ignored attribute intercept with full host visibility");
```
Comment (russell-islam) [inline hypervisor/src/mshv/mod.rs:616]: This should be the case, i.e MSHV is wrongly doing a VM exit. Do you think warn is enough here?

File: hypervisor/src/mshv/mod.rs [hunk 2]
```diff
@@ -606,6 +608,56 @@ impl cpu::Vcpu for MshvVcpu {
...
+                        return Ok(cpu::VmExit::Ignore);
+                    }
+
+                    let num_ranges = info.__bindgen_anon_1.range_count();
+                    assert!(num_ranges >= 1);
+                    if num_ranges > 1 {
+                        warn!("Attribute intercept - expected only 1 gpa range!");
```
Comment (russell-islam) [inline hypervisor/src/mshv/mod.rs:623]: Why one range is a warning? Wording should be more readable like found more than one but expected one only.

File: hypervisor/src/mshv/mod.rs (hunk 1 above)
Comment (jinankjain) [reply hypervisor/src/mshv/mod.rs:616]: I think it is legit to let VM running, if device accidentally give access to same region which it has already provided access to. Ideally it is a bug device side so, I think warning should be good enough here. We should not kill the VM for this reason.

File: hypervisor/src/mshv/mod.rs (hunk 2 above)
Comment (jinankjain) [reply hypervisor/src/mshv/mod.rs:623]: True we should kill the guest here, let me update the code.

File: hypervisor/src/mshv/mod.rs [hunk 3]
```diff
@@ -606,6 +608,59 @@ impl cpu::Vcpu for MshvVcpu {
...
+                    if host_vis >= HV_MAP_GPA_READABLE | HV_MAP_GPA_WRITABLE {
+                        warn!("Ignored attribute intercept with full host visibility");
+                        return Ok(cpu::VmExit::Ignore);
+                    }
+
+                    let num_ranges = info.__bindgen_anon_1.range_count();
+                    assert!(num_ranges >= 1);
+                    if num_ranges > 1 {
+                        return Err(cpu::HypervisorCpuError::RunVcpu(anyhow!(
+                            "Unhandled VCPU exit: Expected num_ranges to be 1 but found num_ranges {:?}",
```
Comment (russell-islam) [inline hypervisor/src/mshv/mod.rs:624]: Something like "Unhandled VCPU exit(GPA_ATTRIBUTE_INTERCEPT ): Expected num_ranges to be 1 but found num_ranges {:?}" makes more sense.

Comment (jinankjain) [reply hypervisor/src/mshv/mod.rs]: Fixed.

**Review by russell-islam — APPROVED** [review]

**Review by likebreath — APPROVED** [review]

**Pull request merged**